### PR TITLE
Add manager classes for platform-specific operations

### PR DESCRIPTION
**Pull request by xas**
## Description

* add `Stm32Manager` class to parse STM32 operations requests from `Options` parameter.
* add `Esp32Manager` class to parse ESP32 operations requests from `Options` parameter.
* add `TIManager` class to parse TI Simplelink operations requests from `Options` parameter.
* add `SilabsManager` class to parse Silabs GG11 operations requests from `Options` parameter.
* add `NanoManager` class to parse nano devices requests from `Options` parameter.
* add `NoOperationPerformedException` for unsuccessful flashing operations

## Motivation and Context

The method `RunOptionsAndReturnExitCodeAsync` is more than 1000 lines of code.  

It includes all the operations for listing, flashing, ... any devices.  

For better readability and maintenance, it could be interesting to separate the operations to a platform-specific manager class.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using System;
+using System.Runtime.Serialization;
+
+namespace nanoFramework.Tools.FirmwareFlasher
+{
+    /// <summary>
+    /// Error executing Silink command.
+    /// </summary>
+    [Serializable]
+    public class SilinkExecutionException : Exception
```
Comment (josesimoes) [inline nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs:15]: There is a typo... I believe you wanted to write `StLink`...

File: nanoFirmwareFlasher.Tool/Esp32Manager.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    /// Class to manage different operations specific to the ESP32 platform.
+    /// </summary>
+    public class Esp32Manager : IManager
+    {
+        private readonly Options _options;
+        private readonly VerbosityLevel _verbosityLevel;
+
+        public Esp32Manager(Options options, VerbosityLevel verbosityLevel)
+        {
+            if (options == null) throw new ArgumentNullException(nameof(options));
```
Comment (josesimoes) [inline nanoFirmwareFlasher.Tool/Esp32Manager.cs:22]: Project code style calls for open and close brackets on IF.

File: nanoFirmwareFlasher.Tool/NanoManager.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace nanoFramework.Tools.FirmwareFlasher
+{
+    /// <summary>
+    /// Class to manage different operations specific to the nano devices.
+    /// </summary>
+    public class NanoManager : IManager
```
Comment (josesimoes) [inline nanoFirmwareFlasher.Tool/NanoManager.cs:13]: Maybe `NanoDeviceManager` ?

**Review by josesimoes — COMMENTED** [review]
Looking good! Just some comments.

File: nanoFirmwareFlasher.Tool/Esp32Manager.cs (hunk 2 above)
Comment (xas) [reply nanoFirmwareFlasher.Tool/Esp32Manager.cs:22]: 👍

File: nanoFirmwareFlasher.Tool/NanoManager.cs (hunk 3 above)
Comment (xas) [reply nanoFirmwareFlasher.Tool/NanoManager.cs:13]: OK for me

File: nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs (hunk 1 above)
Comment (xas) [reply nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs:15]: no, it is about _SI Link_, for the Silabs part. I modify the comment

File: nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        /// <summary>
+        /// SI Link Exception.
+        /// </summary>
+        /// <param name="info">Serialized information.</param>
+        /// <param name="context">Streamed context.</param>
+        protected SilinkExecutionException(SerializationInfo info, StreamingContext context) : base(info, context)
+        {
+        }
+    }
+}
```
Comment (josesimoes) [inline nanoFirmwareFlasher.Library/Exceptions/SilinkExecutionException.cs:56]:
```suggestion
}

```

**Review by josesimoes — APPROVED** [review]
LGTM! Thanks for taking the time to improve this. 👍🏻

**Comment by josesimoes** [comment]
I've tested it with a few different boards and everything seems to be working as expected.

**Pull request merged**